src/core/quantum_gate.jl

```diff
@@ -147,6 +155,145 @@ function apply_gate!(state::Ket, gate::Gate)
+
+    apply_operator!(state,operator,connected_qubits,Int(qubit_count),Val(length(connected_qubits)))
+
+end
+
+get_connected_qubits(gate::AbstractGate)=
+    throw(NotImplementedError(:get_connected_qubits, gate))
+
+
+function apply_operator!(
+    state::Ket,
+    operator::DiagonalOperator,
+    connected_qubit::Int,
+    qubit_count::Int,
+    ::Val{1}) # specialization for single target diagonal gate
+
+    dim=2^qubit_count
+    target_qubit_index=qubit_count-connected_qubit # indexing of targets in qulacs starts at 0
+    
+    diagonal_in_matrix=operator.data
+
+    if target_qubit_index==0
+        for state_index in StepRange(0,2,dim-1)
+            @inbounds state.data[state_index+1] *= diagonal_in_matrix[1];
+            @inbounds state.data[state_index+2] *= diagonal_in_matrix[2];
+        end 
+    else
+        mask = UInt64(1) << target_qubit_index;
+        for state_index in StepRange(0,2,dim-1)
+            bitval = UInt64((state_index & mask) != 0)
+            @inbounds state.data[state_index + 1] *= diagonal_in_matrix[bitval+1];
+            @inbounds state.data[state_index + 2] *= diagonal_in_matrix[bitval+1];
+        end
+    end
+end
+
+# Insert 0 to qubit_index-th bit of basis_index. basis_mask must be 1 << qubit_index.
+function insert_zero_to_basis_index(basis_index::UInt64, basis_mask::UInt64, qubit_index::UInt64)
+    temp_basis = (basis_index >> qubit_index) << (qubit_index + 1)
+    return temp_basis + basis_index % basis_mask
+end
+
+function create_matrix_mask_list(qubit_index_list::Vector{Int64},qubit_index_count::UInt64,matrix_dim::UInt64)
+    mask_list=zeros(UInt64,matrix_dim)
+
+    for cursor in 0:(matrix_dim-1)
+        for bit_cursor in 0:(qubit_index_count-1)
+            if Bool((cursor >> bit_cursor) % 2)
+                bit_index = qubit_index_list[bit_cursor+1]
+                mask_list[cursor+1] ⊻= (UInt64(1) << bit_index) # ⊻ is binary XOR
+            end
+        end
+    end
+
+    return mask_list
+end
+
+function create_shift_mask_list_from_list_buf!(
+    target_qubit_index_list::Vector{Int64},
+    dst_array::Vector{UInt64},
+    dst_mask::Vector{UInt64}
+    )
+    #copy using mutation, not assignment, so dst_array still points to array in caller's scope
+    for (i,target) in enumerate(target_qubit_index_list)
+        dst_array[i]=target 
+    end
+
+    #sort the copy, so the initial array can be used in the original order
+    sort!(dst_array)
+
+    for (i,target) in enumerate(target_qubit_index_list)
+        dst_mask[i]=(1<<target)-1 
+    end
+
+end
+
+function apply_operator!(
+    state::Ket,
+    operator::DiagonalOperator,
+    connected_qubits::Vector{<:Integer},
+    qubit_count::Int,
+    ::Val{N_targets}) where {N_targets} #specialization for N_targets>1 diagonal gates
+
+    dim=2^qubit_count
+    
+    target_qubit_index_list=Vector{Int64}([qubit_count-t for t in reverse(connected_qubits)]) # the bitwise implementation assumes target numbering starting at 0
```

Comment:
This line becomes a bit long with the inline comment. Perhaps we should move it above the line.